extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -6,99 +6,77 @@
 	private _openAsUntitled: boolean;
+	public viewId: string;
+	public books: BookModel[];
+	public currentBook: BookModel;
 
-	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext) {
-		this.initialize(workspaceFolders, null, extensionContext);
+	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext, openAsUntitled: boolean, view: string) {
+		this._openAsUntitled = openAsUntitled;
+		this._extensionContext = extensionContext;
+		this.books = [];
+		this.initialize(workspaceFolders.map(a => a.uri.fsPath));
+		vscode.commands.executeCommand('setContext', 'untitledBooks', openAsUntitled);
+		this.viewId = view;
+		this.prompter = new CodeAdapter();
 	}
 
-	private initialize(workspaceFolders: vscode.WorkspaceFolder[], bookPath: string, context: vscode.ExtensionContext): void {
-		let workspacePaths: string[] = [];
-		if (bookPath) {
-			workspacePaths.push(bookPath);
-		}
-		else if (workspaceFolders) {
-			workspacePaths = workspaceFolders.map(a => a.uri.fsPath);
+	private async initialize(bookPaths: string[]): Promise<any> {
+		if (bookPaths.length > 0) {
+			return Promise.all(bookPaths.map(async (bookPath) => {
+				let book: BookModel = new BookModel(bookPath, this._openAsUntitled, this._extensionContext);
+				await book.initializeContents();
+				this.books.push(book);
+				if (!this.currentBook) {
+					this.currentBook = book;
+				}
+			}));
 		}
-		this.getTableOfContentFiles(workspacePaths).then(() => undefined, (err) => { console.log(err); });
-		this._extensionContext = context;
 	}
 
 	public get onReadAllTOCFiles(): vscode.Event<void> {
 		return this._onReadAllTOCFiles.event;
 	}
 
-	async getTableOfContentFiles(workspacePaths: string[]): Promise<void> {
-		let notebookConfig = vscode.workspace.getConfiguration(notebookConfigKey);
-		let maxDepth = notebookConfig[maxBookSearchDepth];
-		// Use default value if user enters an invalid value
-		if (maxDepth === undefined || maxDepth < 0) {
-			maxDepth = 5;
-		} else if (maxDepth === 0) { // No limit of search depth if user enters 0
-			maxDepth = undefined;
-		}
-		for (let workspacePath of workspacePaths) {
-			let p = path.join(workspacePath, '**', '_data', 'toc.yml').replace(/\\/g, '/');
-			let tableOfContentPaths = await glob(p, { deep: maxDepth });
-			this._tableOfContentPaths = this._tableOfContentPaths.concat(tableOfContentPaths);
-		}
-		let bookOpened: boolean = this._tableOfContentPaths.length > 0;
-		vscode.commands.executeCommand('setContext', 'bookOpened', bookOpened);
-		this._onReadAllTOCFiles.fire();
-	}
 
-	async openBook(bookPath: string, openAsUntitled: boolean, urlToOpen?: string): Promise<void> {
+
+	async openBook(bookPath: string, urlToOpen?: string): Promise<void> {
 		try {
+			let books: BookModel[] = this.books.filter(book => book.bookPath === bookPath);
 			// Check if the book is already open in viewlet.
```

Comment:
We can't assume initialize has been called/is done yet so make sure you check that books is non-undefined (or initialize it to an empty array and make sure it's never set to undefined)